drv/sidecar-front-io/src/transceivers.rs

```diff
@@ -77,6 +90,144 @@ impl FpgaPortMasks {
     }
 }
 
+/// Represents a single logical port (0-31)
+#[derive(Copy, Clone, Debug, PartialEq, PartialOrd)]
+pub struct LogicalPort(pub u8);
+impl LogicalPort {
+    pub fn as_mask(&self) -> LogicalPortMask {
+        LogicalPortMask(1 << self.0)
+    }
+
+    pub fn get_physical_location(&self) -> PortLocation {
+        PORT_MAP[self.0 as usize]
+    }
+}
+/// Represents a set of selected logical ports, i.e. a 32-bit bitmask
+#[derive(Copy, Clone, Debug, Default, PartialEq)]
+pub struct LogicalPortMask(pub u32);
+
+impl LogicalPortMask {
+    pub const MAX_PORT_INDEX: LogicalPort = LogicalPort(NUM_PORTS - 1);
+
+    pub fn get(&self) -> u32 {
+        self.0
+    }
+    pub fn set(&mut self, index: LogicalPort) {
+        self.0 |= index.as_mask().0
+    }
+    pub fn clear(&mut self, index: LogicalPort) {
+        self.0 &= !index.as_mask().0
+    }
+    pub fn merge(&mut self, other: LogicalPortMask) {
+        self.0 |= other.0
+    }
+    pub fn count(&self) -> usize {
+        self.0.count_ones() as _
+    }
+    pub fn is_set(&self, index: LogicalPort) -> bool {
+        self.0 & index.as_mask().0 != 0
+    }
+    pub fn is_empty(&self) -> bool {
+        self.0 == 0
+    }
```

Comment:
I'd suggest rewriting these using the bitwise operations implemented below, e.g.
```suggestion
    pub fn set(&mut self, index: LogicalPort) {
        *self |= index.as_mask()
    }
    pub fn clear(&mut self, index: LogicalPort) {
        *self &= !index.as_mask()
    }
    pub fn merge(&mut self, other: LogicalPortMask) {
        *self |= other
    }
    pub fn count(self) -> usize {
        self.0.count_ones() as _
    }
    pub fn is_set(self, index: LogicalPort) -> bool {
        !(self & index.as_mask()).is_empty()
    }
    pub fn is_empty(self) -> bool {
        self.0 == 0
    }
```
(I also switched from `&self` to just `self` in a few operations, since this object is trivial to clone)